### PR TITLE
Update: support more options in prefer-destructuring

### PR DESCRIPTION
**Pull request by VictorHom**
[X ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))

**What rule do you want to change?**
prefer-destructuring.js
This pr is in reference to the issue #7886 

**Does this change cause the rule to produce more or fewer warnings?**
less warnings since you have more flexibility to control destructuring errors.

**How will the change be implemented? (New option, new default behavior, etc.)?**
The first parameter of destructuring has been updated to also take a VariableDeclarator value or an AssignmentExpression value. Each value takes an object that takes an object of optional keys array and/or object.

The old parameters of  {object: Boolean, array: Boolean} is still supported

**Please provide some example code that this change will affect:**

```json
{
  "rules": {
    "prefer-destructuring": ["error", {
      "VariableDeclaration": {
        "array": false,
        "object": true
      },
     "AssignmentExpression": {
        "array": true,
        "object": true
      }
    }, {
      "enforceForRenamedProperties": false
    }]
  }
}
```

**What does the rule currently do for this code?**
In the above example, prefer-destructuring will be enforced for variable declarations from objects but not from arrays. Prefer-destructuring will be enforced for assignment expressions for both arrays and objects

**What will the rule do after it's changed?**
It will let users have more flexibility and control over when prefer-destructuring should be enforced.

### CODE REVIEW

File: docs/rules/prefer-destructuring.md [hunk 1]
```diff
@@ -33,7 +33,9 @@ This rule takes two sets of configuration objects; the first controls the types
 
 The first has two properties, `array` and `object`, which can be used to turn on or off the destructuring requirement for each of those types independently.  By default, both are `true`.
 
-The second has a single property, `enforceForRenamedProperties`, that controls whether or not the `object` destructuring rules are applied in cases where the variable requires the property being access to be renamed.
+The first has two properties, `VariableDeclarator` and `AssignmentExpression`, which can be used to turn on or off the destructuring requirement for each of those types independently. Each property accepts an object that accpets two properties, `array` and `object`, which can be used to turn on or off the destructuring requirement for each of those types independently for variable declarations and assignment expressions.  By default, `array` and `object` are set to true for both `VariableDeclarator` and `AssignmentExpression`.
```
Comment (not-an-aardvark) [inline docs/rules/prefer-destructuring.md]: Sorry if this was unclear before, but I think we should continue to allow the `array` and `object` top-level options. Making breaking changes to rule schemas is generally a pretty big deal -- in addition to breaking users, it also tends to make shareable configs invalid, which makes upgrading very tedious -- so we try to almost never do it. (Changing the default behavior of a rule schema while keeping all previous schemas valid is a bit better, but that's a separate topic.)

To avoid extra complexity, we can disallow mixing the old and new behaviors -- for example, this would still be an illegal config:

```yml
rules:
  prefer-destructuring: [error, {array: true, VariableDeclarator: {array: false}}]
```

File: docs/rules/prefer-destructuring.md [hunk 2]
```diff
@@ -33,7 +33,9 @@ This rule takes two sets of configuration objects; the first controls the types
...
+
+The second has a single property, `enforceForRenamedProperties`, that controls whether or not the `object` destructuring rules are applied in cases where the variable requires the property being access to be renamed. If `object` in `AssignmentExpression` is set to false, `enforceForRenamedProperties` property is ignored.
```
Comment (not-an-aardvark) [inline docs/rules/prefer-destructuring.md]: Why is `enforceForRenamedProperties` ignored in this case? It seems like it would still be useful in variable declarations for something like

```js
var foo = bar.baz

// ->

var {baz: foo} = bar;
```

File: lib/rules/prefer-destructuring.js [hunk 3]
```diff
@@ -110,22 +136,21 @@ module.exports = {
...
             }
 
-            if (checkObjects && enforceForRenamedProperties) {
+            if ((checkObjectsInVariableDeclarator || checkObjectsInAssignmentExpression) && enforceForRenamedProperties) {
                 report(reportNode, "object");
                 return;
             }
 
-            if (checkObjects) {
+            if ((checkObjectsInVariableDeclarator && reportNode.type === "VariableDeclarator") || (checkObjectsInAssignmentExpression && reportNode.type === "AssignmentExpression")) {
```
Comment (not-an-aardvark) [inline lib/rules/prefer-destructuring.js]: I think the value in this `if` statement is redundant, because `reportNode.type` will always be either `"VariableDeclarator"` or `"AssignmentExpression"`. It can be simplified to

```js
if (checkObjectsInVariableDeclarator || checkObjectsInAssignmentExpression) {
  // ...
}
```

File: lib/rules/prefer-destructuring.js [hunk 4]
```diff
@@ -42,24 +60,32 @@ module.exports = {
...
 
-            if (typeof enabledTypes.object !== "undefined") {
-                checkObjects = enabledTypes.object;
+            if (typeof enabledTypes.AssignmentExpression !== "undefined") {
+                const assignmentExpression = enabledTypes.AssignmentExpression;
+
+                checkArraysInAssignmentExpression = assignmentExpression.hasOwnProperty("array") ? assignmentExpression.array : checkArraysInAssignmentExpression;
+                checkObjectsInAssignmentExpression = assignmentExpression.hasOwnProperty("object") ? assignmentExpression.object : checkObjectsInAssignmentExpression;
             }
         }
```
Comment (not-an-aardvark) [inline lib/rules/prefer-destructuring.js]: This config-checking logic is getting kind of complicated. I think having a variable for each config case won't scale very well as more options are added, especially if we restore support for the old top-level "array" option. Can we extract this into a function instead? For example:

```js
/**
 * @param {string} nodeType "AssignmentExpression" or "VariableDeclaration"
 * @param {string} destructuringType "array" or "object"
 * @returns {boolean} `true` if the destructuring type should be checked for the given node
 */
function shouldCheck(nodeType, destructuringType) {
  return context.options[0] &&
    context.options[0][nodeType] &&
    context.options[0][nodeType][destructuringType] === false;
}
```

Then rather than handling four different variables, the `performCheck` function, it could just use something like

```js
if (isArrayIndexAccess(node) && shouldCheck(reportNode.type, "array")) {
    // ...
}

// ...
```

File: docs/rules/prefer-destructuring.md [hunk 5]
```diff
@@ -62,6 +62,45 @@ An example configuration, with the defaults filled in, looks like this:
...
+var { bar: foo } = object;
+```
+
+An example configuration, with the defaults filled in, looks like this:
+
+```json
+{
+  "rules": {
+    "prefer-destructuring": ["error", {
+      "VariableDeclaration": {
```
Comment (kaicataldo) [inline docs/rules/prefer-destructuring.md]: Should this be `VariableDeclarator`? I actually think `VariableDeclaration` might make more sense.

File: docs/rules/prefer-destructuring.md [hunk 6]
```diff
@@ -62,6 +62,45 @@ An example configuration, with the defaults filled in, looks like this:
 }
 ```
 
+Alternatively, you can use a different set of configuration objects to have more control on what to enforce destructuring. It also takes two sets of configuration objects.
+
+The first has two properties, `VariableDeclarator` and `AssignmentExpression`, which can be used to turn on or off the destructuring requirement for each of those types independently. Each property accepts an object that accpets two properties, `array` and `object`, which can be used to turn on or off the destructuring requirement for each of those types independently for variable declarations and assignment expressions.  By default, `array` and `object` are set to true for both `VariableDeclarator` and `AssignmentExpression`.
```
Comment (kaicataldo) [inline docs/rules/prefer-destructuring.md]: Typo: `accpets` -> `accepts`

File: lib/rules/prefer-destructuring.js [hunk 7]
```diff
@@ -110,22 +167,21 @@ module.exports = {
             }
 
             if (isArrayIndexAccess(rightNode)) {
-                if (checkArrays) {
+                if (shouldCheck("VariableDeclarator", "array") || shouldCheck("AssignmentExpression", "array") || checkArrays) {
```
Comment (kaicataldo) [inline lib/rules/prefer-destructuring.js]: Apologies if I'm misunderstanding something here, but I'm curious what happens for all these checks when the node being evaluated is either a `VariableDeclarator` or `AssignmentExpression` and the configuration is set to `true` for the other node type.

Some example test cases for the above:
```js
{
    code: "foo = object.foo;",
    options: [{ VariableDeclarator: { object: true }, AssignmentExpression: { object: false } }]
},
{
    code: "var foo = object.foo;",
    options: [{ VariableDeclarator: { object: false }, AssignmentExpression: { object: true } }]
}
```

File: docs/rules/prefer-destructuring.md (hunk 5 above)
Comment (VictorHom) [inline docs/rules/prefer-destructuring.md]: just want to make sure - you mean that you would prefer the attribute to be named VariableDeclaration?

File: lib/rules/prefer-destructuring.js (hunk 7 above)
Comment (VictorHom) [inline lib/rules/prefer-destructuring.js]: Good point. With this configuration, both of these are valid cases, along with a few more cases that I added

File: docs/rules/prefer-destructuring.md (hunk 5 above)
Comment (kaicataldo) [inline docs/rules/prefer-destructuring.md]: I feel like it makes more sense, but let's see what others say. I don't want to hold this up over a naming thing, but it is more difficult to change things once they're merged!

Comment (platinumazure) [inline docs/rules/prefer-destructuring.md]: I like VariableDeclarator because I feel it's a closer parallel to AssignmentExpression, but I can see the potential for confusion. No strong opinion from me, though.

Comment (kaicataldo) [inline docs/rules/prefer-destructuring.md]: Let's leave it as VariableDeclarator - I don't want to hold this PR up over a naming thing, especially since I was late to the game in the disxussion. Sorry @VictorHom, looks like you already changed it

Comment (VictorHom) [inline docs/rules/prefer-destructuring.md]: totally fine! I'm sure there's a few more things that could use some cleaning up too.

File: lib/rules/prefer-destructuring.js [hunk 8]
```diff
@@ -56,6 +96,12 @@ module.exports = {
             if (typeof enabledTypes.object !== "undefined") {
                 checkObjects = enabledTypes.object;
             }
+
+            // checkArrays and checkObjects is set to false and use method shouldCheck instead
+            if (typeof enabledTypes.VariableDeclarator !== "undefined" || typeof enabledTypes.AssignmentExpression !== "undefined") {
+                checkArrays = false;
+                checkObjects = false;
+            }
```
Comment (not-an-aardvark) [inline lib/rules/prefer-destructuring.js]: It's not a blocker for this PR, but having two separate methods of checking the options depending on the type of configuration used seems like a bug hazard if something changes later on. Maybe the old config-checking variables could be eliminated too with something like

```js
const normalizedOptions = typeof enabledTypes.array !== "undefined" || typeof context.options[0].object !== "undefined"
    ? { VariableDeclarator: enabledTypes, AssignmentExpression: enabledTypes }
    : enabledTypes
```

In other words, the old `array` and `object` schema would basically be transformed into a special case of the `VariableDeclarator`/`AssignmentExpression` schema where both assignment types happen to have the same configured options.

File: lib/rules/prefer-destructuring.js [hunk 9]
```diff
@@ -110,22 +167,27 @@ module.exports = {
             }
 
             if (isArrayIndexAccess(rightNode)) {
-                if (checkArrays) {
+                if ((shouldCheck("VariableDeclarator", "array") && reportNode.type === "VariableDeclarator") ||
+                    (shouldCheck("AssignmentExpression", "array") && reportNode.type === "AssignmentExpression") ||
```
Comment (not-an-aardvark) [inline lib/rules/prefer-destructuring.js]: I think this could be simplified by just using `if (shouldCheck(reportNode.type, "array"))`.

File: tests/lib/rules/prefer-destructuring.js [hunk 10]
```diff
@@ -35,39 +35,61 @@ ruleTester.run("prefer-destructuring", rule, {
...
             options: [{ object: true }, { enforceForRenamedProperties: false }]
         },
         {
             code: "var foo = object['bar'];",
-            options: [{ object: true }, { enforceForRenamedProperties: false }]
+            options: [{ VariableDeclarator: { object: true } }, { enforceForRenamedProperties: false }]
         },
         {
-            code: "var foo = object[bar];",
+            code: "var foo = object['bar'];",
```
Comment (not-an-aardvark) [inline tests/lib/rules/prefer-destructuring.js]: Is this change necessary? (I don't mind the new version, I'm just sometimes worried about changing old tests because that can prevent regressions from being caught.)

File: tests/lib/rules/prefer-destructuring.js [hunk 11]
```diff
@@ -115,6 +183,14 @@ ruleTester.run("prefer-destructuring", rule, {
...
         {
             code: "var foo = object[bar];",
+            options: [{ VariableDeclarator: { object: true } }, { enforceForRenamedProperties: true }],
+            errors: [{
+                message: "Use object destructuring.",
+                type: "VariableDeclarator"
+            }]
+        },
+        {
+            code: "var foo = object[bar];",
```
Comment (not-an-aardvark) [inline tests/lib/rules/prefer-destructuring.js]: Can you add some additional invalid tests? It would be nice to see:

* A test for the `array` option with the `VariableDeclarator`/`AssignmentExpression` option style
* A test with the `AssignmentExpression` option
* A test where the `VariableDeclarator` and `AssignmentExpression` options are different (e.g. if the code has an object property access without destructuring in an assignment expression, and the options are `{AssignmentExpression: {object: true}, VariableDeclarator: {object: false}}`

Comment (VictorHom) [inline tests/lib/rules/prefer-destructuring.js]: Added more invalid tests

File: tests/lib/rules/prefer-destructuring.js (hunk 10 above)
Comment (VictorHom) [inline tests/lib/rules/prefer-destructuring.js]: changed back. not sure why I updated to add quotes.

File: lib/rules/prefer-destructuring.js (hunk 9 above)
Comment (VictorHom) [inline lib/rules/prefer-destructuring.js]: The main reason why I added the reportNode.type check is to handle cases
```
        {
            code: "foo = array[0];",
            options: [{ VariableDeclarator: { array: true }, AssignmentExpression: { array: false } }, { enforceForRenamedProperties: false }]
        },
        {
            code: "var foo = array[0];",
            options: [{ VariableDeclarator: { array: false }, AssignmentExpression: { array: true } }, { enforceForRenamedProperties: false }]
        },

```
where the check for array/object is swapped between variabledeclarator and assignmentexpression

File: lib/rules/prefer-destructuring.js (hunk 8 above)
Comment (VictorHom) [inline lib/rules/prefer-destructuring.js]: updated to using a normalizedOptions

File: lib/rules/prefer-destructuring.js [hunk 12]
```diff
@@ -15,19 +15,59 @@ module.exports = {
...
+                oneOf: [
+                    {
+                        type: "object",
+                        properties: {
+                            VariableDeclarator: {
+                                type: "object",
+                                properties: {
+                                    array: {
+                                        type: "boolean",
+                                        default: true
```
Comment (VictorHom) [inline lib/rules/prefer-destructuring.js]: I like it for the clarity, but it has no visible effect from running tests.
I used after checking the examples in https://github.com/json-schema-org/json-schema-org.github.io/search?utf8=%E2%9C%93&q=default%3A&type=
let me know if that’s okay. otherwise, I’m not tied to it and if removing makes more sense, then I can remove.

**Comment by VictorHom** [comment]
@not-an-aardvark just want to bring this pr back up for review when you have some time

File: docs/rules/prefer-destructuring.md [hunk 13]
```diff
@@ -29,11 +29,15 @@ var foo = object.bar;
 
 ### Options
 
-This rule takes two sets of configuration objects; the first controls the types that the rule is applied to, and the second controls the way those objects are evaluated.
```
Comment (not-an-aardvark) [inline docs/rules/prefer-destructuring.md]: Can you move this section to before the code samples?

File: lib/rules/prefer-destructuring.js [hunk 14]
```diff
@@ -82,26 +82,18 @@ module.exports = {
     },
     create(context) {
 
-        let checkArrays = true;
-        let checkObjects = true;
-        let enforceForRenamedProperties = false;
         const enabledTypes = context.options[0];
         const additionalOptions = context.options[1];
+        let enforceForRenamedProperties = false;
```
Comment (not-an-aardvark) [inline lib/rules/prefer-destructuring.js]: Nitpick: I think this can just be `let enforceForRenamedProperties = context.options[1] && context.options[1].enforceForRenamedProperties;`

File: lib/rules/prefer-destructuring.js [hunk 15]
```diff
@@ -110,22 +159,26 @@ module.exports = {
             }
 
             if (isArrayIndexAccess(rightNode)) {
-                if (checkArrays) {
+                if ((shouldCheck("VariableDeclarator", "array") && reportNode.type === "VariableDeclarator") ||
+                    (shouldCheck("AssignmentExpression", "array") && reportNode.type === "AssignmentExpression")) {
```
Comment (not-an-aardvark) [inline lib/rules/prefer-destructuring.js]: (Continuing the discussion from https://github.com/eslint/eslint/pull/8796#discussion_r125577727)

I'm not sure I understand -- wouldn't that result in the same behavior? Assuming `reportNode.type` is always either `VariableDeclarator` or `AssignmentExpression`, it seems like `shouldCheck(reportNode.type, "array")` would result in the same behavior. (The `if` statement here has two cases, but only one of them is relevant at any given time, so `shouldCheck(reportNode.type, "array")` would just combine the two cases.)

File: lib/rules/prefer-destructuring.js (hunk 12 above)
Comment (not-an-aardvark) [inline lib/rules/prefer-destructuring.js]: (Continuing the discussion from https://github.com/eslint/eslint/pull/8796#discussion_r125576512)

If it has no effects, then I think we should remove it, because it could be misleading and make people think that it does have an effect. If we want to indicate the default there, I think adding a comment would be better.

File: docs/rules/prefer-destructuring.md [hunk 16]
```diff
@@ -29,11 +29,15 @@ var foo = object.bar;
...
+This rule takes two sets of configuration objects. The first object parameter determines what types of destructuring the rule applies to.
 
-The first has two properties, `array` and `object`, which can be used to turn on or off the destructuring requirement for each of those types independently.  By default, both are `true`.
+The two properties, `array` and `object`, can be used to turn on or off the destructuring requirement for each of those types independently. By default, both are true.
```
Comment (not-an-aardvark) [inline docs/rules/prefer-destructuring.md]: I think it would be useful to give more examples for each option format after these descriptions, e.g.

````markdown
The two properties, `array` and `object`, which can be used to turn on or off the destructuring requirement for each of those types independently. By default, both are true.

For example, the following configuration enforces only object destructuring, but not array destructuring:

```json
{
  "rules": {
    "prefer-destructuring": ["error", {"object": true, "array": false}]
  }
}
```
````

File: docs/rules/prefer-destructuring.md [hunk 17]
```diff
@@ -29,11 +29,15 @@ var foo = object.bar;
...
 
-The second has a single property, `enforceForRenamedProperties`, that controls whether or not the `object` destructuring rules are applied in cases where the variable requires the property being access to be renamed.
+Alternatively, you can use separate configurations for different assignment types. It accepts 2 other keys instead of `array` and `object`.
+
+One key is `VariableDeclarator` and the other is `AssignmentExpression`, which can be used to control the destructuring requirement for each of those types independently. Each property accepts an object that accepts two properties, `array` and `object`, which can be used to control the destructuring requirement for each of `array` and `object` independently for variable declarations and assignment expressions.  By default, `array` and `object` are set to true for both `VariableDeclarator` and `AssignmentExpression`.
```
Comment (not-an-aardvark) [inline docs/rules/prefer-destructuring.md]: Can you add some code samples with the `VariableDeclarator` and `AssignmentExpression` options?

File: lib/rules/prefer-destructuring.js (hunk 12 above)
Comment (kaicataldo) [inline lib/rules/prefer-destructuring.js]: Generally, we've let the documentation and code speak for itself in this area

Comment (VictorHom) [inline lib/rules/prefer-destructuring.js]: I will remove in that case

File: lib/rules/prefer-destructuring.js (hunk 15 above)
Comment (VictorHom) [inline lib/rules/prefer-destructuring.js]: I see what you are saying now. I misinterpreted. sorry for the confusion.

File: docs/rules/prefer-destructuring.md [hunk 18]
```diff
@@ -87,6 +99,42 @@ An example configuration, with the defaults `VariableDeclarator` and `Assignment
...
+
+Examples of **correct** code when object destructuring in `VariableDeclarator` is enforced:
+
+```javascript
+var {bar: foo} = object;
+```
+
+Examples of **correct** code when array destructuring in `AssignmentExpression` is enforced:
+
+```javascript
```
Comment (not-an-aardvark) [inline docs/rules/prefer-destructuring.md]: Can you add a configuration comment to this code block?

```js
/* eslint prefer-destructuring: ["error", {VariableDeclarator: {array: true}}] */
```

**Comment by VictorHom** [comment]
Awesome! thanks for the code reviews.

@kaicataldo can you also give a review as well?
